pkg/controller/jaeger/upgrade.go

```diff
@@ -0,0 +1,25 @@
+package jaeger
+
+import (
+	v1 "github.com/jaegertracing/jaeger-operator/pkg/apis/jaegertracing/v1"
+	"github.com/jaegertracing/jaeger-operator/pkg/upgrade"
+	"github.com/jaegertracing/jaeger-operator/pkg/version"
+)
+
+func (r *ReconcileJaeger) applyUpgrades(jaeger v1.Jaeger) error {
+	currentVersions := version.Get()
+
+	if len(jaeger.Status.Version) > 0 {
+		if jaeger.Status.Version != currentVersions.Jaeger {
```

Comment:
Is that comparison enough to allow update? Based on the operator call we attended the updated should be only allowed if the version in the jaeger spec is "one lower". Eg operator 1.13 can update only Jaeger 1.12.

Reply:
The problem is that may only work if OLM is managing the update, as it will no the previous versions and upgrade one step at a time.

What should happen if the update to the operator is done outside of OLM?

Reply:
If there are 3 new versions in the app registry does OLM deploy sequentially from the older version to the newest one? 

Outside of the OLM users will have to bring sequentially jaeger operator to the version they want.

Reply:
That was my understanding from the meeting - although I think another meeting is planned to discuss upgrades in more detail.

Yes manually will be a pain - plus currently the knowledge of which operator version replaces another is encoded within the csv - so the operator implementation won't necessarily know which version(s) it is replacing to ensure it only tries to update appropriate versions.

Reply:
But maybe that just means upgrades should only be supported when used with OLM.

Reply:
I might be missing something. But the operator should be able to decide whether to update or nor based on the version in the spec and version it supports.

Reply:
Depends what rules we want to follow - for example, if changes are required between two patch versions - how will a manually updated operator know how many patch versions there were between 1.12.1 and 1.13.0?

If the operator implementation maintains update rules, so that it could transition itself from 1.12.1, 1.12.2, 1.12.3 and 1.13.0 that may be one solution  - but I think the presented approach was trying to allow operators to only worry about upgrading a single step.

Reply:
It all depends on the granularity. The version in the spec instance will contain patch version so there are two ways to go
* update if the version difference is `-1` on the patch level.
* update multiple patch levels if our operator maintains multi-patch update
* stop if the version difference is too high.

Reply:
The version in the status field for individual CRs are an indication of the current Jaeger version for that instance. Whenever this is different than the current Operator version, we can assume that an update of the Operator took place, and the operator will run the upgrade routine. A downgrade breaks this assumption, but I don't think we are planning on supporting downgrades.

The way the upgrades work in this PR is via a stepped approach, one upgrade at a time. There's a map with the version string (`v1.13.0`), the function to bring it up from the previous version, and a pointer to the next version. The code for this is in the package `upgrade`. Take a look at `ManagedInstance()` specifically.

OLM might indeed also do a one-at-a-time update, but that won't affect the operator, as the version it will see is just previous+1.

Reply:
I'm marking this conversation (and others) as resolved as I believe this is clarified.